### PR TITLE
Support for dtype

### PR DESCRIPTION
**Pull request by campsd**
Bugfix for #84.

Example code for complex optimization:

```python
import funfact as ff
import numpy as np
ff.use('torch')
m = np.random.rand(4,2,2)
m = m[..., 0] + 1j * m[..., 1]
n = np.random.rand(2,3,2)
n = n[..., 0] + 1j * n[..., 1]
target = m @ n
a = ff.tensor(4,2)
b = ff.tensor(2,3)
i,j,k = ff.indices(3)
tsrex = a[i,j] * b[j,k]
fac = ff.factorize(tsrex, target, nvec=64, max_steps=10000, dtype='target')
```

### CODE REVIEW

File: funfact/algorithm.py [hunk 1]
```diff
@@ -65,7 +73,11 @@ def factorize(
             that represents all the solutions.
     '''
 
-    target = ab.tensor(target)
+    if dtype == 'target':
+        dtype = str(target.dtype)
+        dtype = operator.attrgetter(dtype)(ab)
```
Comment (yhtang) [inline funfact/algorithm.py:78]: I suppose this is, for example, to convert numpy dtypes to the equivalent ones in the backend?

How about flipping the order of this `if` block with line 80, so that we can directly use the type of the new `target` tensor?

File: funfact/backend/__init__.py [hunk 2]
```diff
@@ -55,6 +59,8 @@ def use(backend: str):
     '''
     global _active_backend
     try:
+        if enable_x64 and backend == 'jax':
+            os.environ['JAX_ENABLE_X64'] = 'True'
```
Comment (yhtang) [inline funfact/backend/__init__.py:63]: Can we also verify if x64 is enabled using the method as suggested by the JAX [docs](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html):
> We can then confirm that x64 mode is enabled:
``` py
import jax.numpy as jnp
from jax import random
x = random.uniform(random.PRNGKey(0), (1000,), dtype=jnp.float64)
x.dtype # --> dtype('float64')
```

File: funfact/backend/_torch.py [hunk 3]
```diff
@@ -20,11 +20,11 @@ def tensor(cls, array, optimizable=False, **kwargs):
         return torch.tensor(array, requires_grad=optimizable, **kwargs)
 
     @classmethod
-    def to_numpy(cls, tensor):
+    def to_numpy(cls, tensor, **kwargs):
         if tensor.requires_grad:
-            return tensor.detach().numpy()
+            return np.asarray(tensor.detach().numpy(), **kwargs)
```
Comment (yhtang) [inline funfact/backend/_torch.py:25]: This can be made shorter by
```
if tensor.requires_grad:
    tensor = tensor.detach()
return np.asarray(...)
```

File: funfact/loss.py [hunk 4]
```diff
@@ -45,7 +45,7 @@ def __call__(self, model, target, reduction='mean', sum_vec=True,
 class MSE(Loss):
 
     def _loss(self, model, target):
-        return ab.square(ab.subtract(model, target))
+        return ab.square(ab.abs(ab.subtract(model, target)))
```
Comment (yhtang) [inline funfact/loss.py:48]: Maybe leave a comment to say that `abs` is for handling complex data? Otherwise, it may appear as redundant given the `square` operation.

File: funfact/algorithm.py [hunk 5]
```diff
@@ -12,7 +13,7 @@
 def factorize(
     tsrex, target, lr=0.1, tol=1e-6, max_steps=10000, optimizer='Adam',
     loss='mse_loss', nvec=1, stop_by='first', returns='best',
-    checkpoint_freq=50, **kwargs
+    checkpoint_freq=50, dtype='target', **kwargs
```
Comment (yhtang) [inline funfact/algorithm.py:16]: How about `dtype=None`, and then infer `dtype` if it is None? This should be as intuitive as `'target'`, and since infer dtype from target is the only special case here, we don't really need to use a string to name it.

File: funfact/algorithm.py [hunk 6]
```diff
@@ -55,6 +56,13 @@ def factorize(
             - If 'all', return all instances.
 
         checkpoint_freq (int >= 1): The frequency of convergence checking.
+
+        dtype: The datatype of the factorization model ('target', ab.dtype):
+
+            - If 'target', the same data type as the target tensor is used.
```
Comment (yhtang) [inline funfact/algorithm.py:62]: If `None`

**Review by yhtang — CHANGES_REQUESTED** [review]

File: funfact/backend/__init__.py (hunk 2 above)
Comment (campsd) [reply funfact/backend/__init__.py]: I have added this to `_jax.py`. However, this warning will only trigger if JAX was already loaded separately from FunFact before with a different X64 setting. E.g.
```
import jax # load 32 bit
...
ff.use('jax', enable_x64=True) # warning will be shown here
```

 If the user does something like:
```
ff.use('jax', enable_x64=False) # load 32 bit
...
ff.use('jax', enable_x64=True) # try to load 64 bit
```
The second run does not actually get imported again presumably because in
```
_active_backend = getattr(
            importlib.import_module(f'funfact.backend._{backend}'), clsname
        )
```
`_jax.py` has already been loaded before. So the warning is not thrown in this case.

File: funfact/backend/_torch.py (hunk 3 above)
Comment (campsd) [reply funfact/backend/_torch.py]: Done.

File: funfact/loss.py (hunk 4 above)
Comment (campsd) [reply funfact/loss.py:49]: Done.

File: funfact/algorithm.py (hunk 6 above)
Comment (campsd) [reply funfact/algorithm.py]: Done

File: funfact/algorithm.py (hunk 1 above)
Comment (campsd) [reply funfact/algorithm.py:78]: The idea by first converting it to a native type tensor was that `dtype = str(target.dtype)` always gives a dtype from the backend. 

For example with torch this will become something like `'torch.float32'` which can be found by `operator.attrgetter`. If we flip it around, we can get something like `'numpy.float3'` which would cause an error in line 78 as it is not an attr of `ab`.

File: funfact/algorithm.py (hunk 5 above)
Comment (campsd) [reply funfact/algorithm.py]: Done

File: funfact/algorithm.py (hunk 1 above)
Comment (yhtang) [reply funfact/algorithm.py:78]: Oh, I meant what if we do this:
``` py
if dtype is None:
    target = ab.tensor(target)
    dtype = target.dtype
else:
    target = ab.tensor(target, dtype=dtype)
    dtype = target.dtype
```
and then be done. If the user wanna explicit control of dtype, they will manually specify a dtype from `ab`, or convert target to `ab.native_t` before passing it into vectorize.

File: funfact/backend/__init__.py (hunk 2 above)
Comment (yhtang) [reply funfact/backend/__init__.py]: I see. Sounds really tricky. How about we just stay as-is and assume that users will do things in the right order for the time being then? :cry:

**Review by yhtang — APPROVED** [review]

**Comment by yhtang** [comment]
Could you also please check on the failed tests? :smile:

File: funfact/backend/__init__.py (hunk 2 above)
Comment (campsd) [reply funfact/backend/__init__.py]: Ok, I have removed this again.

**Comment by campsd** [comment]
The tests are working now.

I'll look into that JAX complex conjugate issue in a separate PR.

**Review by yhtang — APPROVED** [review]

**Comment by yhtang** [comment]
> 

Cool. Thanks!
